src/main/java/net/neoforged/neoforge/registries/RegistryManager.java

```diff
@@ -24,158 +22,245 @@
-    private final String name;
+    /**
+     * Schedule a registry to be registered to {@link BuiltInRegistries#REGISTRY}.
+     * Called from {@link RegistryBuilder#create()}. Do not try to call this directly.
+     */
+    synchronized static void addRegistry(Registry<?> registry) {
+        if (pendingRegistries == null) {
+            throw new IllegalStateException("Too late to add registries, register them in your mod constructor!");
+        }
 
-    RegistryManager() {
-        this("STAGING");
-    }
+        ResourceLocation registryName = registry.key().location();
 
-    public RegistryManager(String name) {
-        this.name = name;
-    }
+        if (BuiltInRegistries.REGISTRY.containsKey(registryName) || pendingRegistries.stream().anyMatch(reg -> reg.key() == registry.key()))
+            throw new IllegalStateException("Cannot create a registry that already exists - " + registryName);
 
-    public String getName() {
-        return this.name;
+        pendingRegistries.add(registry);
     }
 
-    boolean isStaging() {
-        return "STAGING".equals(this.name);
-    }
+    public synchronized static void createRegistries() {
+        List<Registry<?>> moddedStaticRegistries = pendingRegistries;
+        pendingRegistries = null; // Prevent further registration
 
-    @SuppressWarnings("unchecked")
-    public <V> ForgeRegistry<V> getRegistry(ResourceLocation key) {
-        return (ForgeRegistry<V>) this.registries.get(key);
-    }
+        DataPackRegistryEvent.NewRegistry dataPackEvent = new DataPackRegistryEvent.NewRegistry();
+        vanillaRegistryKeys = Set.copyOf(BuiltInRegistries.REGISTRY.keySet());
 
-    public <V> ForgeRegistry<V> getRegistry(ResourceKey<? extends Registry<V>> key) {
-        return getRegistry(key.location());
-    }
+        ModLoader.get().postEventWrapContainerInModOrder(dataPackEvent);
+
+        registerModdedRegistries(moddedStaticRegistries);
+        dataPackEvent.process();
 
-    public <V> ResourceLocation getName(IForgeRegistry<V> reg) {
-        return this.registries.inverse().get(reg);
+        BuiltInRegistries.REGISTRY.forEach(RegistryManager::postModifyRegistryEvent);
     }
 
-    public <V> ResourceLocation updateLegacyName(ResourceLocation legacyName) {
-        ResourceLocation originalName = legacyName;
-        while (getRegistry(legacyName) == null) {
-            legacyName = legacyNames.get(legacyName);
-            if (legacyName == null) {
-                return originalName;
+    private static void registerModdedRegistries(List<Registry<?>> registries) {
+        RuntimeException aggregate = new RuntimeException();
+
+        ((BaseMappedRegistry<?>) BuiltInRegistries.REGISTRY).unfreeze();
+
+        for (final var registry : registries) {
+            try {
+                registerToRootRegistry(registry);
+            } catch (Throwable t) {
+                aggregate.addSuppressed(t);
+                return;
```

Comment:
Shouldn't this just `throw` instead of bothering with an aggregate that is not even used?

Reply:
Looks like the intention was to collect all of these errors and then log the aggregate below but the return statement prevents that. In my opinion, either the aggregate or the exception itself should definitely be thrown though and not just logged

Reply:
I chose to let the exception be thrown as that's the easiest solution.